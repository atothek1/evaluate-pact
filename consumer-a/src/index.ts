import { get, post } from "superagent";
import { Product } from "./types";

export class ProductApi {
    constructor(private baseUrl: string) {
        this.baseUrl = `${baseUrl}/v1/products`
    }
    async getProducts() {
        const response = await get(this.baseUrl);
        return response.body as Promise<Product[]>;
    }
    async addProduct(product: Product) {
        return await post(this.baseUrl)
            .send(product)
            .then(response => response.body)
            .catch(err => {throw err.response.body})
    }
}

export async function getProducts(baseUrl: string): Promise<Product[]> {
    const url = baseUrl + "/v1/products";
    const response = await get(url);
    return response.body as Promise<Product[]>;
}