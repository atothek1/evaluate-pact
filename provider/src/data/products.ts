const faker = require("faker")
import { Product } from "../../src/types";

const { commerce, datatype, image } = faker;
faker.seed(5678);

function createImageUrls(): string[] {
    const res: string[] = [];
    for (let i = 0; i < 10; i++) {
        res.push(image.imageUrl());
    }
    return res
}

export function createProduct(): Product {
    return {
        id: datatype.uuid(),
        name: commerce.productName(),
        price: parseFloat(commerce.price()) * 10,
        description: commerce.productDescription(),
        isAvailable: datatype.boolean(),
        imageUrls: createImageUrls()
    }
}
export class Products {
    constructor(private products: Product[] = []) { }
    clear() { this.products = [] }
    add(...products: Product[]) { this.products.push(...products) }
    get() { return [ ...this.products ] }
}
export const productRepository = new Products();