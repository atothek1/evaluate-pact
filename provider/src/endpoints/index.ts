import { Matchers, RequestOptions, ResponseOptions } from '@pact-foundation/pact';
import { createProduct, productRepository } from '../data/products';
import { Product } from '../types';

const { like } = Matchers;
function createProducts(count: number = 10): Product[] {
    const res: Product[] = [];
    for (let i = 0; i < count; i++) {
        res.push(createProduct());
    }
    return res
}

export const productsRequest : { uponReceiving: string, withRequest: RequestOptions } = {
    uponReceiving: 'a request for fetching all products',
    withRequest: {
        method: 'GET',
        path: '/v1/products',
    },
};

export const withListOfProductsRequest : {state: string | undefined, willRespondWith: ResponseOptions } = {
    state: "has a collection of products",
    willRespondWith: {
        status: 200,
        headers: {
            "Content-Type": "application/json; charset=utf-8",
        },
        body: like(createProducts()),
    }
};