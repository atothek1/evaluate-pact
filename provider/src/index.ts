import express from "express";
import { createProduct, productRepository } from "./data/products";
import { Product } from "./types";

export const app = express();

app.use((_req, res, next) => {
    res.header("Content-Type", "application/json; charset=utf-8");
    next();
})
app.use(express.json())

const tryCreateProduct = async (body: any): Promise<Product> => {
    // okay, not a real promise yet. but possibly will be
    return new Promise((resolve, reject) => {
        if (!body) { return reject("Missing body") }
        try {
            const productTemplate = createProduct()
            // we don't have a database yet, so we generate an id manually
            body.id = productTemplate.id;
            const productDefinition = Object.keys(productTemplate)
            resolve(productDefinition
                .reduce((result, field) => {
                    if (!body[field]) throw `Missing field '${field}' in payload`
                    result[field] = body[field]
                    return result
                }, {} as any) as Product)
        } catch (e) {
            console.error(e)
            reject(e)
        }
    })
}

app.route("/v1/products")
    .get((req, res) => {
        const { isAvailable } = req.query;
        const result = isAvailable !== undefined
            ? productRepository.get().filter(p => p.isAvailable)
            : productRepository.get();
        if (isAvailable)
            console.log("Return the following results", result, "Based on the filter", isAvailable)
        res.status(200).send(result);
    })
    .post((req, res) => {
        const body = req.body as Product;
        tryCreateProduct(body)
            .then(product => (productRepository.add(product), res.status(201).send({ ...product })))
            .catch(e => res.status(400).send({ success: false, error: e }))
    });

app.get("/v1/products/:id", (req, res) => {
    const { id } = req.params;
    const result = productRepository.get().find(p => p.id === id);
    if (result) {
        res.status(200).send(result);
    } else {
        res.status(404).send("not found")
    }
});

export const server = app.listen(3001, () => {
    console.log("pact evaluation provider running: http://localhost:3001")
})