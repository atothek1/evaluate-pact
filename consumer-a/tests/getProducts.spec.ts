import { pactWith } from 'jest-pact';
import { ProductApi } from "../src";
import { productsRequest, withListOfProductsRequest } from "provider/src/endpoints";
import { Matchers } from '@pact-foundation/pact';
import { createProduct } from 'provider/src/data/products';
import { Product } from '../src/types';
import { addProduct, failProductCreation, succeedProductCreation } from './productRequests';

const { like } = Matchers;

pactWith({ consumer: 'e2e Consumer A', provider: 'e2e Provider' }, provider => {
  let client: ProductApi;

  beforeEach(async () => {
    await provider.removeInteractions()
    client = new ProductApi(provider.mockService.baseUrl)
  });

  describe('getProducts()', () => {

    test('if products are available, it returns a list of products', async () =>{      
      await provider.addInteraction({
        ...productsRequest,
        ...withListOfProductsRequest
      })
      await client.getProducts().then(products => {
        expect(products).toMatchSnapshot()
      })
    });

    test('if no products are available, it returns an empty list', async () =>{      
      await provider.addInteraction({
        ...productsRequest,
        state: "has no products",
        willRespondWith: {
            status: 200,
            headers: {
                "Content-Type": "application/json; charset=utf-8",
            },
            body: like([]),
        }
      })
      await client.getProducts().then(products => {
        expect(products).toMatchSnapshot()
      })
    });
  });

  test('adding a product fails if required field is missing', async () => {
      const product = {
        ...createProduct(),
        description: undefined
      } as any as Product;
      
      await provider.addInteraction({
        ...addProduct(product),
        ...failProductCreation('fails if required field is missing', "Missing field 'description' in payload")
      })
      await client.addProduct(product as any as Product).catch(error => {
        expect(error).toMatchSnapshot()
      })
      expect.assertions(1)
    });
    
   test('adding a product succeeds for a valid product', async () => {
      const product = createProduct() as any as Product;
      await provider.addInteraction({
        ...addProduct(product),
        ...succeedProductCreation('succeeds for a valid product', product)
      })
      await client.addProduct(product).then(product => {
        expect(product).toMatchSnapshot()
      })
      expect.assertions(1)
    });
})