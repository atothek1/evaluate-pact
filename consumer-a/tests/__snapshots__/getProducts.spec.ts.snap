// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pact between e2e Consumer A and e2e Provider with 30000 ms timeout for Pact adding a product fails if required field is missing 1`] = `
Object {
  "error": "Missing field 'description' in payload",
}
`;

exports[`Pact between e2e Consumer A and e2e Provider with 30000 ms timeout for Pact adding a product succeeds for a valid product 1`] = `
Object {
  "description": "The automobile layout consists of a front-engine design, with transaxle-type transmissions mounted at the rear of the engine and four wheel drive",
  "id": "d724e8bc-94c2-4841-9a66-579e91110ca3",
  "imageUrls": Array [
    "http://placeimg.com/640/480",
    "http://placeimg.com/640/480",
    "http://placeimg.com/640/480",
    "http://placeimg.com/640/480",
    "http://placeimg.com/640/480",
    "http://placeimg.com/640/480",
    "http://placeimg.com/640/480",
    "http://placeimg.com/640/480",
    "http://placeimg.com/640/480",
    "http://placeimg.com/640/480",
  ],
  "isAvailable": true,
  "name": "Generic Steel Shoes",
  "price": 6950,
}
`;

exports[`Pact between e2e Consumer A and e2e Provider with 30000 ms timeout for Pact getProducts() if no products are available, it returns an empty list 1`] = `Array []`;

exports[`Pact between e2e Consumer A and e2e Provider with 30000 ms timeout for Pact getProducts() if products are available, it returns a list of products 1`] = `
Array [
  Object {
    "description": "Boston's most advanced compression wear technology increases muscle oxygenation, stabilizes active muscles",
    "id": "780a5887-9d6e-42b4-8192-f14bc32c7a2c",
    "imageUrls": Array [
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
    ],
    "isAvailable": true,
    "name": "Sleek Soft Chips",
    "price": 4170,
  },
  Object {
    "description": "The slim & simple Maple Gaming Keyboard from Dev Byte comes with a sleek body and 7- Color RGB LED Back-lighting for smart functionality",
    "id": "b1df0207-6e6c-4d02-87af-f9e39646a121",
    "imageUrls": Array [
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
    ],
    "isAvailable": false,
    "name": "Licensed Plastic Tuna",
    "price": 8440,
  },
  Object {
    "description": "Boston's most advanced compression wear technology increases muscle oxygenation, stabilizes active muscles",
    "id": "24a24975-7863-4e71-958c-8dafe3e50906",
    "imageUrls": Array [
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
    ],
    "isAvailable": false,
    "name": "Tasty Cotton Pants",
    "price": 9930,
  },
  Object {
    "description": "The slim & simple Maple Gaming Keyboard from Dev Byte comes with a sleek body and 7- Color RGB LED Back-lighting for smart functionality",
    "id": "8ed1e4bc-b41e-4b82-a04b-b8eddc85172c",
    "imageUrls": Array [
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
    ],
    "isAvailable": false,
    "name": "Generic Frozen Gloves",
    "price": 5770,
  },
  Object {
    "description": "New range of formal shirts are designed keeping you in mind. With fits and styling that will make you stand apart",
    "id": "68912a13-0c07-4cac-801d-693b1ad7b90b",
    "imageUrls": Array [
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
    ],
    "isAvailable": false,
    "name": "Small Metal Mouse",
    "price": 9510,
  },
  Object {
    "description": "Ergonomic executive chair upholstered in bonded black leather and PVC padded seat and back for all-day comfort and support",
    "id": "ffe42eba-b20c-4d19-a2e5-929e365f2697",
    "imageUrls": Array [
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
    ],
    "isAvailable": false,
    "name": "Intelligent Wooden Soap",
    "price": 2570,
  },
  Object {
    "description": "The Apollotech B340 is an affordable wireless mouse with reliable connectivity, 12 months battery life and modern design",
    "id": "83346bf2-e9c8-4d3d-a66d-19a95dce6d60",
    "imageUrls": Array [
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
    ],
    "isAvailable": true,
    "name": "Awesome Frozen Chair",
    "price": 400,
  },
  Object {
    "description": "The automobile layout consists of a front-engine design, with transaxle-type transmissions mounted at the rear of the engine and four wheel drive",
    "id": "88bae26a-f4ad-42e3-9f06-5d5e3b14ae15",
    "imageUrls": Array [
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
    ],
    "isAvailable": true,
    "name": "Awesome Granite Gloves",
    "price": 7260,
  },
  Object {
    "description": "The beautiful range of Apple Natural√© that has an exciting mix of natural ingredients. With the Goodness of 100% Natural Ingredients",
    "id": "54efd241-92b0-45fe-a24d-7975646f4aba",
    "imageUrls": Array [
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
    ],
    "isAvailable": true,
    "name": "Small Fresh Pizza",
    "price": 2930,
  },
  Object {
    "description": "The Apollotech B340 is an affordable wireless mouse with reliable connectivity, 12 months battery life and modern design",
    "id": "5e7d496f-0190-4b09-bee8-efcf3113495e",
    "imageUrls": Array [
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
    ],
    "isAvailable": false,
    "name": "Incredible Soft Bacon",
    "price": 7120,
  },
]
`;
